File: solutions/chapter4/problem_4_2.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    int N;
    cin >> N;
    vector<int> memo(N+1);

    for (int i = 0; i <= N; i++) {
        if (i <= 1) {
            memo[i] = 0;
            continue;
        } else if (i == 2) {
            memo[i] = 1;
            continue;
        }
        memo[i] = memo[i-1]+memo[i-2]+memo[i-3];
    }

    cout << memo[N] << endl;
}

File: solutions/chapter4/problem_4_2.py
def main():
    N = int(input())
    memo = [0] * (N+1)
    
    for i in range(N+1):
        if i <= 1:
            continue
        if i == 2:
            memo[i] = 1
            continue
        memo[i] = memo[i-1]+memo[i-2]+memo[i-3]
    
    print(memo[N])

if __name__ == '__main__':
    main()